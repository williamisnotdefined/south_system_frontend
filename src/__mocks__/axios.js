const mockAxios = jest.genMockFromModule('axios')
mockAxios.create = jest.fn(() => mockAxios)

mockAxios.get = async param => {
    if (param) {
        return Promise.resolve({
            id: 0,
            name: 'Green Dragon',
            type: 'Green',
            createdAt: '2020-05-24T09:45:33.595Z'
        })
    }

    return Promise.resolve([
        {
            id: 0,
            name: 'Green Dragon',
            type: 'Green',
            createdAt: '2020-05-24T09:45:33.595Z'
        },
        {
            id: 1,
            name: 'Dragon Lord',
            type: 'Red',
            createdAt: '2020-05-24T09:45:33.595Z'
        }
    ])
}

mockAxios.post = async (q, { name, type }) =>
    Promise.resolve({
        id: 2,
        name,
        type,
        createdAt: '2020-05-24T09:45:33.595Z'
    })

mockAxios.delete = async ({ name, type }) =>
    Promise.resolve({
        id: 1,
        name,
        type,
        createdAt: '2020-05-24T09:45:33.595Z'
    })

mockAxios.put = async ({ name, type }) =>
    Promise.resolve({
        id: 1,
        name,
        type,
        createdAt: '2020-05-24T09:45:33.595Z'
    })

export default mockAxios
