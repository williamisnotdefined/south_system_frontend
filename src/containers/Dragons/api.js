import api from '@services/api'

import { Creators } from './index'
import * as toastify from '@services/toastify'

export async function loadDragons(dispatch) {
    dispatch(Creators.loadingDragons())

    try {
        const dragons = await api.get()
        dispatch(Creators.loadedDragons(dragons))
    } catch (err) {
        toastify.error('Error on load dragons.', 3000)
    }
}

export async function deleteDragon(dispatch, id) {
    toastify.info('Removing dragon..')

    try {
        await api.delete(`/${id}`)
        dispatch(Creators.removeDragon(id))
        toastify.clear()
        toastify.success('Dragon removed successfully.')
    } catch (err) {
        toastify.clear()
        toastify.error('Error on delete dragon.', 3000)
    }
}
