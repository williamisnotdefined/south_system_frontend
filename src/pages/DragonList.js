import React from 'react'

export default function DragonList() {
    return <h1>Dragon List</h1>
}
