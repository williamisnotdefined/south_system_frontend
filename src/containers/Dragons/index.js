import React, { createContext, useReducer, useCallback } from 'react'
import PropTypes from 'prop-types'

import sortDragons from '@helpers/sortDragon'

import * as dragonRequest from './api'

const initialState = {
    dragons: [],
    loadingList: false,
    saving: false
}

const DragonContext = createContext(initialState)

const Actions = {
    LOAD_DRAGONS: '@Dragons/LOAD_DRAGONS',
    LOADED_DRAGONS: '@Dragons/LOADED_DRAGONS',

    ADD_DRAGON: '@Dragons/ADD_DRAGON',
    REMOVE_DRAGON: '@Dragons/REMOVE_DRAGON',

    SET_SAVING_DRAGON: '@Dragons/SET_SAVING_DRAGON',
    UNSET_SAVING_DRAGON: '@Dragons/UNSET_SAVING_DRAGON'
}

const Creators = {
    loadingDragons: () => ({ type: Actions.LOAD_DRAGONS }),
    loadedDragons: dragons => ({ type: Actions.LOADED_DRAGONS, payload: dragons }),

    removeDragon: dragonId => ({ type: Actions.REMOVE_DRAGON, payload: dragonId }),
    addDragon: dragon => ({ type: Actions.ADD_DRAGON, payload: dragon }),

    setSavingDragon: () => ({ type: Actions.SET_SAVING_DRAGON }),
    unsetSavingDragon: () => ({ type: Actions.UNSET_SAVING_DRAGON })
}

const reducer = (state, { type, payload }) => {
    switch (type) {
        case Actions.LOAD_DRAGONS:
            return { ...state, loadingList: true }
        case Actions.LOADED_DRAGONS:
            const sortedDragons = sortDragons(payload)
            return { ...state, dragons: sortedDragons, loadingList: false }

        case Actions.ADD_DRAGON:
            const allDragons = [...state.dragons, payload]
            const sortedDragonsAdded = sortDragons(allDragons)
            return { ...state, dragons: sortedDragonsAdded, saving: false }

        case Actions.REMOVE_DRAGON:
            const dragons = state.dragons.filter(dragon => dragon.id !== payload)
            return { ...state, dragons }

        case Actions.SET_SAVING_DRAGON:
            return { ...state, saving: true }
        case Actions.UNSET_SAVING_DRAGON:
            return { ...state, saving: false }

        default:
            return { ...state }
    }
}

function DragonsProvider({ children }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    const loadDragons = useCallback(() => dragonRequest.loadDragons(dispatch), [])
    const deleteDragon = useCallback(dragonId => dragonRequest.deleteDragon(dispatch, dragonId), [])
    const addDragon = useCallback(dragon => dragonRequest.addDragon(dispatch, dragon), [])

    return (
        <DragonContext.Provider value={{ state, dispatch, loadDragons, deleteDragon, addDragon }}>
            {children}
        </DragonContext.Provider>
    )
}

DragonsProvider.propTypes = {
    children: PropTypes.node
}

export { Creators, DragonContext, DragonsProvider }
