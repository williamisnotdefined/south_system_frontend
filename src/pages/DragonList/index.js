import React, { useContext, useEffect } from 'react'

import { DragonContext } from '@containers/Dragons'

import PageTitle from '@components/PageTitle'
import Loader from '@components/Loader'
import DragonCard from '@components/DragonCard'

import { Container, ListWrapper, LoaderWrapper } from './styles'

export default function DragonList() {
    const { state, loadDragons } = useContext(DragonContext)
    const { dragons, loadingList } = state

    useEffect(() => {
        loadDragons()
    }, [loadDragons])

    return (
        <Container>
            <PageTitle>Dragon List</PageTitle>
            <ListWrapper>
                {loadingList && (
                    <LoaderWrapper>
                        <Loader />
                    </LoaderWrapper>
                )}
                {!loadingList &&
                    dragons.map((dragon, key) => <DragonCard key={key} dragon={dragon} />)}
            </ListWrapper>
        </Container>
    )
}
