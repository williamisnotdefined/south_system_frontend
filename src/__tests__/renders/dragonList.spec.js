import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

import { theme } from '@styles'
import { DragonContext, initialState } from '@containers/Dragons'
import DragonList from '@pages/DragonList'

import axios from 'axios'

jest.mock('axios')

describe('Testing Dragon List Page Render', () => {
    it('should render correctly', async () => {
        const apiDragons = await axios.get()

        const contextValue = { state: { ...initialState, dragons: apiDragons, firstLoad: false } }
        const list = renderer
            .create(
                <ThemeProvider theme={theme}>
                    <DragonContext.Provider value={contextValue}>
                        <DragonList />
                    </DragonContext.Provider>
                </ThemeProvider>
            )
            .toJSON()

        expect(list).toMatchSnapshot()
    })
})
