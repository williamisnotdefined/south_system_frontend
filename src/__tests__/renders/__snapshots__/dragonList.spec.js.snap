// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Dragon List Page Render should render correctly 1`] = `
<div
  className="sc-AxmLO fOmAtm"
>
  <h1
    className="sc-AxjAm gwODik"
  >
    Dragon List
  </h1>
  <ul
    className="sc-fzozJi ghudOy"
  >
    <li
      className="sc-AxhCb gfatYW"
      onClick={[Function]}
    >
      <img
        className="sc-AxhUy ZofuJ"
        src="dragon_lord.gif"
      />
      <p
        className="sc-AxgMl cpNlvo"
      >
        Green Dragon
      </p>
      <p
        className="sc-AxheI cYcWID"
      >
        Green
      </p>
      <button
        className="sc-Axmtr dZrOXu"
        onClick={[Function]}
      >
        X
      </button>
    </li>
    <li
      className="sc-AxhCb gfatYW"
      onClick={[Function]}
    >
      <img
        className="sc-AxhUy ZofuJ"
        src="dragon_lord.gif"
      />
      <p
        className="sc-AxgMl cpNlvo"
      >
        Dragon Lord
      </p>
      <p
        className="sc-AxheI cYcWID"
      >
        Red
      </p>
      <button
        className="sc-Axmtr dZrOXu"
        onClick={[Function]}
      >
        X
      </button>
    </li>
  </ul>
</div>
`;
